### ISSUE TITLE
Darktable 4.4.2 Freezing on multiple operations

### ISSUE THREAD

**Issue by MarkKocherovsky** [darktable-org/darktable#15234]
### Describe the bug

I effectively cannot use Darktable as when I try to do several operations --- including simply zooming in --- darktable hangs and freezes, requiring me to forcibly restart my computer.

### Steps to reproduce

Open Darktable
Move to the darkroom
Zoom in

### Expected behavior

Zoom in/perform complex operations

### Where did you install darktable from?

darktable.org

### darktable version

4.4.2

### What OS are you using?

Linux

### What is the version of your OS?

Ubuntu 22.04.3

### Describe your system?

CPU: AMD Ryzen 7 6800H with Radeon Graphics
Memory: 16 GB
Swap space: 16 GB
GPU: NVIDIA GeForce RTX 3060 Mobile

### Are you using OpenCL GPU in darktable?

Yes

### If yes, what is the GPU card and driver?

NVIDIA GeForce RTX 3060 Mobile, Driver version 525

### Please provide additional context if applicable. You can attach files too, but might need to rename to .txt or .zip

The following are my processing preferences:

![image](https://github.com/darktable-org/darktable/assets/67556945/38f9bde9-d08b-4f67-8740-e552cbb9bf83)

**Comment by MarkKocherovsky** [comment]
Ah of course now that I post the issue it decides to work, I'll leave this open for the moment in case anyone has any suggestions for optimizing it.

**Comment by jenshannoschwalm** [comment]
You probably want to set
1. the tune cl preferences to default. transfer makes no sense for nvidia, memory size might be the culprit if you are using firefox for example as that might take more cl memory than wanted.
2. prefer performance over quality is not recommended for such fast machines as it severely impairs quality.

**Comment by gi-man** [comment]
Also, you seem to have two GPU. AMD could cause issues depending on the drivers. I think you should disable the AMD. In 4.4.2 you need to edit darktablerc. In 4.6 (or current master), it is a checkbox.

**Comment by MarkKocherovsky** [comment]
> 

This did not help. Specifically, it hung once I went to RGB Color Balance, selected the Path mask option, and zoomed in to make a path

**Comment by jenshannoschwalm** [comment]
1. Sorry, if you want further help we would need a debug logfile with `-d pipe -d opencl` :-)
2. Also you might want to check the manual, as suggested above use default settings.
3. AMD drivers are notorious for problems, so a log file will keep you on track.

There are "hundreds" of issues around - here and in forums - related to wrong OpenCL installations, Ubuntu 22 has been problematic for a while. Many issues have now been fixed in dt master, you might give it a try :-)


Also, please keep in mind, this is an issue tracker, you might be better off in pixels.us, there are lots of friendly people helping out with such cases and you will also see threads about problems like this.

**Comment by MarkKocherovsky** [comment]
I've gotten

```
     0.0465 [dt_get_sysresource_level] switched to 1 as `default'
     0.0465   total mem:       15162MB
     0.0465   mipmap cache:    1895MB
     0.0465   available mem:   7581MB
     0.0465   singlebuff:      118MB
     0.0465   OpenCL tune mem: OFF
     0.0465   OpenCL pinned:   OFF
[opencl_init] opencl related configuration options:
[opencl_init] opencl: ON
[opencl_init] opencl_scheduling_profile: 'very fast GPU'
[opencl_init] opencl_library: 'default path'
[opencl_init] opencl_device_priority: '*/!0,*/*/*/!0,*'
[opencl_init] opencl_mandatory_timeout: 20000
     0.0472 [dt_dlopencl_init] could not find default opencl runtime library 'libOpenCL'
     0.0473 [dt_dlopencl_init] could not find default opencl runtime library 'libOpenCL.so'
[opencl_init] opencl library 'libOpenCL.so.1' found on your system and loaded
[opencl_init] found 1 platform
[opencl_init] found 1 device

[dt_opencl_device_init]
   DEVICE:                   0: 'NVIDIA GeForce RTX 3060 Laptop GPU'
   PLATFORM NAME & VENDOR:   NVIDIA CUDA, NVIDIA Corporation
   CANONICAL NAME:           nvidiacudanvidiageforcertx3060laptopgpu
   DRIVER VERSION:           525.125.06
   DEVICE VERSION:           OpenCL 3.0 CUDA, SM_20 SUPPORT
   DEVICE_TYPE:              GPU
   GLOBAL MEM SIZE:          5938 MB
   MAX MEM ALLOC:            1484 MB
   MAX IMAGE SIZE:           32768 x 32768
   MAX WORK GROUP SIZE:      1024
   MAX WORK ITEM DIMENSIONS: 3
   MAX WORK ITEM SIZES:      [ 1024 1024 64 ]
   ASYNC PIXELPIPE:          NO
   PINNED MEMORY TRANSFER:   NO
   MEMORY TUNING:            NO
   FORCED HEADROOM:          400
   AVOID ATOMICS:            NO
   MICRO NAP:                250
   ROUNDUP WIDTH:            16
   ROUNDUP HEIGHT:           16
   CHECK EVENT HANDLES:      128
   PERFORMANCE:              3.365
   TILING ADVANTAGE:         0.000
   DEFAULT DEVICE:           NO
   KERNEL BUILD DIRECTORY:   /opt/darktable/share/darktable/kernels
   KERNEL DIRECTORY:         /home/markkocherovsky/.cache/darktable/cached_v1_kernels_for_NVIDIACUDANVIDIAGeForceRTX3060LaptopGPU_52512506
   CL COMPILER OPTION:       -cl-fast-relaxed-math
   KERNEL LOADING TIME:       0.0170 sec
[opencl_init] OpenCL successfully initialized. Internal numbers and names of available devices:
[opencl_init]		0	'NVIDIA CUDA NVIDIA GeForce RTX 3060 Laptop GPU'
[opencl_init] FINALLY: opencl is AVAILABLE and ENABLED.
[dt_opencl_update_priorities] these are your device priorities:
[dt_opencl_update_priorities] 		image	preview	export	thumbs	preview2
[dt_opencl_update_priorities]		0	0	0	0	0
[dt_opencl_update_priorities] show if opencl use is mandatory for a given pixelpipe:
[dt_opencl_update_priorities] 		image	preview	export	thumbs	preview2
[dt_opencl_update_priorities]		1	1	1	1	1
[opencl_synchronization_timeout] synchronization timeout set to 0
[dt_opencl_update_priorities] these are your device priorities:
[dt_opencl_update_priorities] 		image	preview	export	thumbs	preview2
[dt_opencl_update_priorities]		0	0	0	0	0
[dt_opencl_update_priorities] show if opencl use is mandatory for a given pixelpipe:
[dt_opencl_update_priorities] 		image	preview	export	thumbs	preview2
[dt_opencl_update_priorities]		1	1	1	1	1
[opencl_synchronization_timeout] synchronization timeout set to 0
```

**Comment by MarkKocherovsky** [comment]
I've gotten

```
     0.0465 [dt_get_sysresource_level] switched to 1 as `default'
     0.0465   total mem:       15162MB
     0.0465   mipmap cache:    1895MB
     0.0465   available mem:   7581MB
     0.0465   singlebuff:      118MB
     0.0465   OpenCL tune mem: OFF
     0.0465   OpenCL pinned:   OFF
[opencl_init] opencl related configuration options:
[opencl_init] opencl: ON
[opencl_init] opencl_scheduling_profile: 'very fast GPU'
[opencl_init] opencl_library: 'default path'
[opencl_init] opencl_device_priority: '*/!0,*/*/*/!0,*'
[opencl_init] opencl_mandatory_timeout: 20000
     0.0472 [dt_dlopencl_init] could not find default opencl runtime library 'libOpenCL'
     0.0473 [dt_dlopencl_init] could not find default opencl runtime library 'libOpenCL.so'
[opencl_init] opencl library 'libOpenCL.so.1' found on your system and loaded
[opencl_init] found 1 platform
[opencl_init] found 1 device

[dt_opencl_device_init]
   DEVICE:                   0: 'NVIDIA GeForce RTX 3060 Laptop GPU'
   PLATFORM NAME & VENDOR:   NVIDIA CUDA, NVIDIA Corporation
   CANONICAL NAME:           nvidiacudanvidiageforcertx3060laptopgpu
   DRIVER VERSION:           525.125.06
   DEVICE VERSION:           OpenCL 3.0 CUDA, SM_20 SUPPORT
   DEVICE_TYPE:              GPU
   GLOBAL MEM SIZE:          5938 MB
   MAX MEM ALLOC:            1484 MB
   MAX IMAGE SIZE:           32768 x 32768
   MAX WORK GROUP SIZE:      1024
   MAX WORK ITEM DIMENSIONS: 3
   MAX WORK ITEM SIZES:      [ 1024 1024 64 ]
   ASYNC PIXELPIPE:          NO
   PINNED MEMORY TRANSFER:   NO
   MEMORY TUNING:            NO
   FORCED HEADROOM:          400
   AVOID ATOMICS:            NO
   MICRO NAP:                250
   ROUNDUP WIDTH:            16
   ROUNDUP HEIGHT:           16
   CHECK EVENT HANDLES:      128
   PERFORMANCE:              3.365
   TILING ADVANTAGE:         0.000
   DEFAULT DEVICE:           NO
   KERNEL BUILD DIRECTORY:   /opt/darktable/share/darktable/kernels
   KERNEL DIRECTORY:         /home/markkocherovsky/.cache/darktable/cached_v1_kernels_for_NVIDIACUDANVIDIAGeForceRTX3060LaptopGPU_52512506
   CL COMPILER OPTION:       -cl-fast-relaxed-math
   KERNEL LOADING TIME:       0.0170 sec
[opencl_init] OpenCL successfully initialized. Internal numbers and names of available devices:
[opencl_init]		0	'NVIDIA CUDA NVIDIA GeForce RTX 3060 Laptop GPU'
[opencl_init] FINALLY: opencl is AVAILABLE and ENABLED.
[dt_opencl_update_priorities] these are your device priorities:
[dt_opencl_update_priorities] 		image	preview	export	thumbs	preview2
[dt_opencl_update_priorities]		0	0	0	0	0
[dt_opencl_update_priorities] show if opencl use is mandatory for a given pixelpipe:
[dt_opencl_update_priorities] 		image	preview	export	thumbs	preview2
[dt_opencl_update_priorities]		1	1	1	1	1
[opencl_synchronization_timeout] synchronization timeout set to 0
[dt_opencl_update_priorities] these are your device priorities:
[dt_opencl_update_priorities] 		image	preview	export	thumbs	preview2
[dt_opencl_update_priorities]		0	0	0	0	0
[dt_opencl_update_priorities] show if opencl use is mandatory for a given pixelpipe:
[dt_opencl_update_priorities] 		image	preview	export	thumbs	preview2
[dt_opencl_update_priorities]		1	1	1	1	1
[opencl_synchronization_timeout] synchronization timeout set to 0
```
I'll try concurrently updating to the master branch if that would help.

**Comment by MarkKocherovsky** [comment]
Is cloning into ``/darktable/src/tests/integration/`` supposed to take a long time?

**Comment by jenshannoschwalm** [comment]
Looks good. Now do the Log while you run into issue 😀

**Comment by MarkKocherovsky** [comment]
Alright, I've updated to **4.5.0+593~g7b7272a11a** and the log as it freezes is attached as an image:

![image](https://github.com/darktable-org/darktable/assets/67556945/825f916e-1a92-40cc-bbf4-b031b5dde866)

**Comment by gi-man** [comment]
Can you post the log in txt file or pastebin or something similar that allows reading multiple rows. 

Are your raw files hosted locally?

**Comment by MarkKocherovsky** [comment]
> Can you post the log in txt file or pastebin or something similar that allows reading multiple rows.

Where, approximately are the logs located?

> Are your raw files hosted locally?

Yes, they're Canon CR3 raw files, if that helps

**Comment by aagit** [comment]
I might be reproducing a similar opencl stall with 4.4.2 too, but only after zooming in on rocm-5.4.2 self built with ebuild (couldn't make 5.7.1 not to segfault) on a amd mobile apu gfx90c, which I wasn't even sure if it could run rocm on a CPU with only 15W TDP. Initially it appeared the GPU was slower than the CPU because of the stall.

The pipeline gets unstuck immediately after I remove focus from the darktable window, like moving the mouse to a different window or if I watch a youtube video on firefox in another plasma screen. So I wonder if it might be perhaps related to kwin_wayland. I haven't tested it on x11.

video stopped:

```
     4.9241 [dev_pixelpipe] took 0.010 secs (0.000 CPU) [preview] processed `demosaic' on GPU, blended on GPU
     4.9587 [dev_pixelpipe] took 0.035 secs (0.002 CPU) [preview] processed `denoiseprofile' on GPU, blended on GPU
--
     6.5138 [dev_pixelpipe] took 0.199 secs (0.003 CPU) [full] processed `demosaic' on GPU, blended on GPU
     6.7290 [dev_pixelpipe] took 0.215 secs (0.003 CPU) [full] processed `denoiseprofile' on GPU, blended on GPU
--
     8.9640 [dev_pixelpipe] took 0.115 secs (0.007 CPU) [full] processed `demosaic' on GPU, blended on GPU
    10.2470 [dev_pixelpipe] took 1.283 secs (0.009 CPU) [full] processed `denoiseprofile' on GPU, blended on GPU
     ^^^^^^
--
    12.2937 [dev_pixelpipe] took 0.117 secs (0.117 CPU) [full] processed `demosaic' on GPU, blended on GPU
    12.3559 [dev_pixelpipe] took 0.062 secs (0.064 CPU) [full] processed `denoiseprofile' on GPU, blended on GPU
--
    14.4063 [dev_pixelpipe] took 0.056 secs (0.004 CPU) [full] processed `demosaic' on GPU, blended on GPU
    16.3724 [dev_pixelpipe] took 1.966 secs (0.012 CPU) [full] processed `denoiseprofile' on GPU, blended on GPU
    ^^^^^^^
```

video playing:

```
     5.7228 [dev_pixelpipe] took 0.238 secs (0.018 CPU) [full] processed `demosaic' on GPU, blended on GPU
     5.7826 [dev_pixelpipe] took 0.060 secs (0.006 CPU) [full] processed `denoiseprofile' on GPU, blended on GPU
--
     6.8845 [dev_pixelpipe] took 0.010 secs (0.000 CPU) [preview] processed `demosaic' on GPU, blended on GPU
     6.9214 [dev_pixelpipe] took 0.037 secs (0.008 CPU) [preview] processed `denoiseprofile' on GPU, blended on GPU
--
     8.7807 [dev_pixelpipe] took 0.239 secs (0.180 CPU) [full] processed `demosaic' on GPU, blended on GPU
     8.8290 [dev_pixelpipe] took 0.048 secs (0.005 CPU) [full] processed `denoiseprofile' on GPU, blended on GPU
--
    10.0234 [dev_pixelpipe] took 0.112 secs (0.032 CPU) [full] processed `demosaic' on GPU, blended on GPU
    10.0761 [dev_pixelpipe] took 0.053 secs (0.002 CPU) [full] processed `denoiseprofile' on GPU, blended on GPU
--
    11.6982 [dev_pixelpipe] took 0.103 secs (0.098 CPU) [full] processed `demosaic' on GPU, blended on GPU
    11.7618 [dev_pixelpipe] took 0.064 secs (0.041 CPU) [full] processed `denoiseprofile' on GPU, blended on GPU
--
    12.8591 [dev_pixelpipe] took 0.048 secs (0.027 CPU) [full] processed `demosaic' on GPU, blended on GPU
    12.9106 [dev_pixelpipe] took 0.051 secs (0.006 CPU) [full] processed `denoiseprofile' on GPU, blended on GPU
--
    14.5780 [dev_pixelpipe] took 0.051 secs (0.052 CPU) [full] processed `demosaic' on GPU, blended on GPU
    14.6322 [dev_pixelpipe] took 0.054 secs (0.051 CPU) [full] processed `denoiseprofile' on GPU, blended on GPU
--
    15.7084 [dev_pixelpipe] took 0.026 secs (0.001 CPU) [full] processed `demosaic' on GPU, blended on GPU
    15.7666 [dev_pixelpipe] took 0.058 secs (0.002 CPU) [full] processed `denoiseprofile' on GPU, blended on GPU
```

No idea if it's the same issue, the system remains stable, it continues the pipeline after stalling for some second, but I thought I shall report it.

Thank you.

**Comment by gi-man** [comment]
I think rocm does not support iGPU. Can you reproduce without using rocm? Can you reproduce with current master? If yes, start a new issue.

**Comment by aagit** [comment]
> I think rocm does not support iGPU. Can you reproduce without using rocm? Can you reproduce with current master? If yes, start a new issue.

What I reproduced seems a different issue and not related to darktable. I I tested darktable latest git on rocm 5.5.1 and 5.4.3 (5.7.1 keeps segfaulting). The stall always reproduces even if less frequently on 5.5.1. Looking at https://github.com/ROCm/ROCm/issues/2675 it starts to look a rocm problem not related to darktable but reporting much detail there may not lead to much activity if the APU is unsupported as you mentioned.  It feels like there could be an easy fix though even if it might not come from ROCm upstream.

I also see lots of mbind to node 1 around the GPU alocations. This CPU has two NUMA nodes with node 0 having 16 threads, and node 1 with zero threads. I probably should try to boot with NUMA disabled in case the CPU-less node where the APU pci resides is confusing rocm libs doing the memory bindings. I don't think discrete CPUs ever run on systems with CPU-less NUMA nodes. And then I may try again with x11. All the stall time is spent calling futex and getclock, there's no GPU ioctl activity during the stall.

I think you can close this for now sorry for the noise.

**Comment by jenshannoschwalm** [comment]
Thanks for the testing.

**Comment by aagit** [comment]
update: I might have found a workaround, still using rocm 5.5.1 darktable 4.4.2 on 7730U gfx90c. Setting CHECK EVENT HANDLES (f. number of event handles) to 1 might have resolved my rendering lag.

cldevice_v5_amdacceleratedparallelprocessinggfx90cxnack=0 0 0 16 16 1 1 0 0.059351 0.000

I also enabled async to 1 and dropped the micro nap to zero because it appeared to run a bit faster  (about 10%) that way with no apparent adverse effect.

So now I could compare more reliably the performance of rocm vs CPU. rocm opencl:

    [dev_process_image] pixel pipeline took 0.340 secs (1.385 CPU)

CPU only:

    [dev_process_image] pixel pipeline took 0.552 secs (5.226 CPU)

So rocm is about 60% faster at full zoom in, at max zoom out it's less noticeable but still about 30% faster.

I also noticed I was seeing overflows in the output of darktable -d perf, probably because you're using the tsc directly. If you used one of the nsec gettimeofday variants it wouldn't be much slower if the tsc is in sync thanks to the vsyscall. I figured this laptop is affected by a widespread BIOS bug that caused a time warp of the tsc on CPU0 and prevented the tsc clocksource and the vsyscalls to be used. So I also worked around that to get a stable tsc clocksource using a tsc=directsync out of tree kernel patch. Not sure if that also contributed to make rocm work more stable (if not rock solid), probably not.
